### ISSUE TITLE
Row height for componentRenderer not rendering correctly

### ISSUE THREAD

**Issue by StefanCov** [vaadin/vaadin-grid#1684]
I have created tiles for the grid which display on a single row but having a few problems with the row height not rendering properly. This seem's to be at random occurrences and changing the browser width or refreshing changes the problem (Sometimes fixes the row height). I don't know if this is caused by the large height of the component or slowness of the rendering which affects the display. I have no idea how to fix this or force the row to be 100% of the containing component.

This is currently what is happening to my components within the grid.
![image](https://user-images.githubusercontent.com/54804121/71973894-05ffd180-3208-11ea-82fa-970fb6745dff.png)

I have hid the details within the component but the green line shows how the component should end. The red line shows that the buttons have been cut off way before the full size of the component. This can cause the bottom buttons to be cut off all together.

Any ideas for a fix?

**Comment by StefanCov** [comment]
Any update on this? I found clicking on the row forces the rows to size correctly. Think this is due to the way it shows the row selected.

**Comment by StefanCov** [comment]
Is there a way to possibly programmatically highlight the row like it has been clicked on after refreshing the data?

**Comment by TatuLund** [comment]
Are you referring the selection?

**Comment by StefanCov** [comment]
I managed to get the first row to be selected which works as a temporary fix as it resizes all the rows once one row selected.

`grid.getSelectionModel().select(dataProviderList.get(0));` 

This was the code which selects my first row but it is not ideal.

**Comment by StefanCov** [comment]
Sadly selecting the first row has not fixed the problem.

**Comment by hongduc-phan** [comment]
@StefanCov : Can you add the simple code to reproduce the bug here? Thank you

**Comment by tomivirkki** [comment]
Repro: https://pastebin.com/5BAv6bbE

**Comment by StefanCov** [comment]
These are the tile components which I place 3 within a FlexLayout. The Flexlayout is then added to the Grid.

**Comment by Haprog** [comment]
@StefanCov I tried reproducing the issue by taking the above provided code for `<product-tile>` and by making a grid like this:
```html
<vaadin-grid id="grid">
  <vaadin-grid-column>
    <template>
      <div style="display: flex;">
        <product-tile></product-tile>
        <product-tile></product-tile>
        <product-tile></product-tile>
      </div>
    </template>
  </vaadin-grid-column>
</vaadin-grid>
```

And setting few items (just to render few rows):
```js
ready() {
    super.ready();
    this.$.grid.items = [1, 2, 3, 4];
}
```

I added all the imports needed for the used components:
```html
<link rel="import" href="../../bower_components/polymer/polymer-element.html">
<link rel="import" href="../../bower_components/vaadin-icons/vaadin-icons.html">
<link rel="import" href="../../bower_components/vaadin-button/vaadin-button.html">
<link rel="import" href="../../bower_components/vaadin-ordered-layout/vaadin-horizontal-layout.html">
<link rel="import" href="../../bower_components/vaadin-dialog/vaadin-dialog.html">
<link rel="import" href="../../bower_components/vaadin-grid/vaadin-grid.html">
<link rel="import" href="../../bower_components/vaadin-grid/vaadin-grid-column.html">
```

The code for `<quantity-selector>` was not provided so that's not imported, but it probably doesn't affect the issue.

Since you mention `FlexLayout` I assume you're using also Flow where you create and populate the grid, but I wanted to try reproduce this simply without Flow to make things easier to debug.

I'm not sure how exactly you're adding the contents to the grid or what content is left out from above code, but I think the problem probably is just that within the `<product-tile>` elements there's some content (for example images) that gets loaded after the `<product-tile>` has otherwise been rendered and when that content loads it affects the height of the tile, but at that point the grid row height doesn't update anymore since it has been already rendered and the contents may overflow the row height.

On my test the grid looks like this:
![image](https://user-images.githubusercontent.com/75155/76752073-7c7cfa80-6788-11ea-8bd0-d2e1b4656752.png)

I tried to simulate dynamically loaded content by adding this to product-tile class:
```js
ready() {
    super.ready();
    setTimeout(() => {
        this.$.imageDiv.style.height = '70px';
    }, 190);
}
```

Which results in:
![image](https://user-images.githubusercontent.com/75155/76752809-f7deac00-6788-11ea-9a6a-9368fa804e4c.png)

When this happens it can be fixed by manually triggering a rerender of the grid by calling `grid.render()` on the `<vaadin-grid>` element. You can try this in the browser dev tools by selecting the `<vaadin-grid>` element and running `$0.render();`. See https://vaadin.com/components/vaadin-grid/html-api/elements/Vaadin.GridElement#method-render

So if you know when all the content that may affect row height have been loaded you can manually run this. You could trigger it from Flow with `grid.getElement().executeJs("this.render();");`

In case of images being loaded, you could add a `load` event listener on the image elements to know when an image has loaded but it can be complex to try to track loading states of all the images in the grid which have been added to DOM and only after all of them have loaded, then call `render()`.

The easiest solution would be to make sure with CSS that the row height doesn't change. For example having images with a fixed height or only in containers with fixed height.

I'm making some assumptions here since I wasn't given a full working app that reproduces the exact issue you have.

Does this help?

If you think you have some other issue, I would need more details (preferably a working project with minimal code to reproduce the issue) to be able to reproduce it and investigate further.

**Comment by StefanCov** [comment]
Hi, sorry I couldn't provide much more evidence as i'd need to simplify the code that I had and I'm currently not working on this project at the moment. I think you have nailed the problem on the head though. It must have something to do with the images or icons I load on the tile which change in size. When I get more time to work on my Vaadin project, I will try add this render fix as I should know when the grid has rendered.

**Comment by tomivirkki** [comment]
Note: `grid.getElement().executeJs("this.notifyResize();")` should be more lightweight than `grid.getElement().executeJs("this.render();")` as it doesn't re-invoke the renderers but just re-measures the existing content.

**Comment by Haprog** [comment]
Since it seems this is only about the content height being updated dynamically after render, I will close this as invalid since it's not a bug.

Hopefully when you get back to your project you can make it work using either `notifyResize()`, `render()` or adding some CSS styles for fixing the content height. But if not and you think it is another issue than described above, then please open a new issue about it with more details about how to reproduce it so we can take a closer look.
